File: ts/day_6/main.ts
export type Input = Fish;

export function readData(): Input {
  const contents = Deno.readTextFileSync("./data/day_6.txt").trim();

  return parseContents(contents);
}

export function parseContents(contents: string): Input {
  return contents.split(",").reduce((acc, num) => {
    acc[+num]++;
    return acc;
  }, [0, 0, 0, 0, 0, 0, 0, 0, 0] as Fish);
}

export function part1(data: Input): number {
  return fishAfterDays(data, 80);
}

export function part2(data: Input): number {
  return fishAfterDays(data, 256);
}

function fishAfterDays(data: Fish, days: number): number {
  for (let i = 0; i < days; i++) {
    data.push(data.shift() as number);
    data[6] += data[8];
  }
  return data.reduce((acc, val) => acc + val);
}

export type Fish = [
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
  number,
];

export function main() {
  const data = readData();

  console.log("Part 1: ", part1(data));
  console.log("Part 2: ", part2(data));
}

File: rs/day_6/src/main.rs
#![deny(clippy::all)]
#![feature(test)]
extern crate test;

use std::{fs::read_to_string, path::Path};

type Input = [usize; 9];

fn main() {
    let data = read_data("./data");

    println!("Part 1: {}", part_1(&data));
    println!("Part 2: {}", part_2(&data));
}

fn read_data(data_dir: &str) -> Input {
    let contents = read_to_string(Path::new(data_dir).join("day_6.txt")).unwrap();

    parse_contents(contents.trim())
}

fn parse_contents(contents: &str) -> Input {
    contents.split(',').fold([0usize; 9], |mut acc, num| {
        acc[num.parse::<usize>().unwrap()] += 1;
        acc
    })
}

fn part_1(input: &Input) -> usize {
    fish_after_days(input, 80)
}

fn part_2(input: &Input) -> usize {
    fish_after_days(input, 256)
}

fn fish_after_days(input: &Input, days: usize) -> usize {
    (0..days)
        .fold(*input, |mut fish, _| {
            fish.rotate_left(1);
            fish[6] += fish[8];
            fish
        })
        .iter()
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;
    use test::Bencher;

    const PART_1: usize = 360761;
    const PART_2: usize = 1632779838045;

    mod read_data {
        use super::*;

        #[bench]
        fn actual(b: &mut Bencher) {
            b.iter(|| {
                let data = read_data("../../data");

                assert_ne!(data, Input::default())
            })
        }
    }

    mod parse_contents {
        use super::*;

        struct Case<'c> {
            input: &'c str,
            expected: Input,
        }

        #[test]
        fn example() {
            run(&Case {
                input: "3,4,3,1,2",
                expected: example_data(),
            })
        }

        fn run(test: &Case) {
            assert_eq!(test.expected, parse_contents(test.input))
        }
    }

    mod part_1 {
        use super::*;

        struct Case {
            data: Input,
            expected: usize,
        }

        #[test]
        fn example() {
            run(&Case {
                data: example_data(),
                expected: 5934,
            })
        }

        #[bench]
        fn actual(b: &mut Bencher) {
            let case = Case {
                data: read_data("../../data"),
                expected: PART_1,
            };

            b.iter(|| run(&case))
        }

        fn run(test: &Case) {
            assert_eq!(test.expected, part_1(&test.data))
        }
    }

    mod part_2 {
        use super::*;

        struct Case {
            data: Input,
            expected: usize,
        }

        #[test]
        fn example() {
            run(&Case {
                data: example_data(),
                expected: 26984457539,
            })
        }

        #[bench]
        fn actual(b: &mut Bencher) {
            let case = Case {
                data: read_data("../../data"),
                expected: PART_2,
            };

            b.iter(|| run(&case))
        }

        fn run(test: &Case) {
            assert_eq!(test.expected, part_2(&test.data))
        }
    }

    fn example_data() -> Input {
        [0, 1, 1, 2, 1, 0, 0, 0, 0]
    }
}
